synfig-core/src/synfig/os.cpp

```diff
@@ -0,0 +1,632 @@
+			NULL,          // use parent's environment
+			NULL,          // use parent's current directory
+			&siStartInfo,  // STARTUPINFO pointer
+			&piProcInfo);  // receives PROCESS_INFORMATION
+
+		// If an error occurs, exit the application.
+		if (!bSuccess) {
+		   synfig::error("synfig::OS::pipe: CreateProcess");
+		   return false;
+		}
+		// Close handles to the child process and its primary thread.
+		// Some applications might keep these handles to monitor the status
+		// of the child process, for example.
+
+		CloseHandle(piProcInfo.hProcess);
+		CloseHandle(piProcInfo.hThread);
+
+		// Close handles to the stdin and stdout pipes no longer needed by the child process.
+		// If they are not explicitly closed, there is no way to recognize that the child process has ended.
+
+		CloseHandle(child_STDOUT_Write);
+		child_STDOUT_Write = INVALID_HANDLE_VALUE;
+		CloseHandle(child_STDIN_Read);
+		child_STDIN_Read = INVALID_HANDLE_VALUE;
+
+		return true;
+	}
+};
+#else
+class RunPipeUnix : public OS::RunPipe
+{
+	FILE* read_file = nullptr;
+	FILE* write_file = nullptr;
+	bool initialized = false;
+	pid_t pid = -1;
+
+	// RunPipe interface
+public:
+	RunPipeUnix() = default;
+	~RunPipeUnix()
+	{
+		RunPipeUnix::close();
+	}
+
+	bool open(std::string binary_path, const OS::RunArgs& binary_args, OS::RunMode run_mode, const std::string& stdout_redir_file="", const std::string& stdin_redir_file="") override
+	{
+		if (initialized) {
+			synfig::error("A pipe should not be initialized twice! Ignored...");
+			return false;
+		}
+		initialized = true;
+
+		full_command_ = binary_path + " " + binary_args.get_string();
+
+		int p[2];
+
+		if (pipe(p)) {
+			synfig::error(_("Unable to open pipe to %s (no pipe)"), binary_path.c_str());
+			return false;
+		}
+
+		pid = fork();
+
+		if (pid == -1) {
+			synfig::error(_("Unable to open pipe to %s (pid == -1)"), binary_path.c_str());
+			return false;
+		}
+		const bool is_reader = run_mode == OS::RUN_MODE_READ || run_mode == OS::RUN_MODE_READWRITE;
+		const bool is_writer = run_mode == OS::RUN_MODE_WRITE || run_mode == OS::RUN_MODE_READWRITE;
+
+		if (pid == 0) {
+			// Child process
+			// Redirect stdout to writer-pipe or to a file
+			if (is_reader || !stdout_redir_file.empty()) {
+				if (!stdout_redir_file.empty()) {
+					::close(p[1]);
+					mode_t mode = S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH;
+					p[1] = ::open(stdout_redir_file.c_str(), O_WRONLY | O_CREAT | O_TRUNC, mode);
+					if (p[1] == -1) {
+						synfig::error(_("Unable to open file to be stdout: %s [errno: %d]"), stdout_redir_file.c_str(), errno);
+						perror(_("System error message:"));
+						exit(-1000);
+					}
+				}
+				// Map pipeout as stdout
+				if (dup2(p[1], STDOUT_FILENO) == -1) {
+					synfig::error(_("Unable to open pipe to %s (dup2( p[1], STDOUT_FILENO ) == -1)"), binary_path.c_str());
+					exit(-2000);
+				}
+				// Map stderr to pipeout as well
+				if (dup2(p[1], STDERR_FILENO) == -1) {
```

Comment:
I realized it's not correct to merge stdout and stderr into one single pipe (or redirection file).
FFmpeg always output its "about message" via stderr and it prevents to import a file o.O

Should I change the `open()` signature to add a new parameter?
`bool open(std::string binary_path, const OS::RunArgs& binary_args, OS::RunMode run_mode, const std::string& stdout_redir_file="", const std::string& stdin_redir_file="") override`

it's so loooong

Reply:
I think better to do this.
You can check `Glib::spawn_sync` for ideas.